File: Cryambly/Engine/Logic/Slots/EntitySlotInfo.cs
using System;
using System.Runtime.InteropServices;
using CryCil.Annotations;
using CryCil.Engine.Models.StaticObjects;
using CryCil.Engine.Rendering;
using CryCil.Engine.Rendering.Lighting;

namespace CryCil.Engine.Logic
{
	/// <summary>
	/// Encapsulates information about a slot.
	/// </summary>
	[StructLayout(LayoutKind.Sequential)]
	public unsafe struct EntitySlotInfo
	{
		#region Fields
		[UsedImplicitly] private EntitySlotFlags nFlags;
		[UsedImplicitly] private int nParentSlot;
		// Hide mask used by breakable object to indicate what index of the CStatObj sub-object is hidden.
		[UsedImplicitly] private ulong nSubObjHideMask;
		[UsedImplicitly] private Matrix34* pLocalTM;
		[UsedImplicitly] private Matrix34* pWorldTM;
		[UsedImplicitly] private EntityId entityId;
		[UsedImplicitly] private IntPtr pStatObj;
		[UsedImplicitly] private IntPtr pCharacter;
		[UsedImplicitly] private IntPtr pParticleEmitter;
		[UsedImplicitly] private IntPtr pLight;
		[UsedImplicitly] private IntPtr pChildRenderNode;
		[UsedImplicitly] private IntPtr pGeomCacheRenderNode;
		[UsedImplicitly] private Material pMaterial;
		#endregion
		#region Properties
		/// <summary>
		/// Gets a set of flags that describe this slot.
		/// </summary>
		public EntitySlotFlags Flags
		{
			get { return this.nFlags; }
		}
		/// <summary>
		/// Gets a zero-based index of the entity slot that is set as a parent for this one. Returns -1 if
		/// no parent is set.
		/// </summary>
		public int ParentSlot
		{
			get { return this.nParentSlot; }
		}
		/// <summary>
		/// Gets a 3x4 matrix that represents local transformation of this slot.
		/// </summary>
		public Matrix34 LocalTransformationMatrix
		{
			get { return *this.pLocalTM; }
		}
		/// <summary>
		/// Gets a 3x4 matrix that represents world transformation of this slot.
		/// </summary>
		public Matrix34 WorldTransformationMatrix
		{
			get { return *this.pWorldTM; }
		}
		/// <summary>
		/// Gets identifier of the entity that is bound to this slot (Returns invalid identifier, if no
		/// entity is bound.)
		/// </summary>
		public EntityId BoundEntity
		{
			get { return this.entityId; }
		}
		/// <summary>
		/// Gets valid object that represents the particle emitter that is bound to this slot. Returned
		/// object is not valid if the emitter is not bound to this slot.
		/// </summary>
		public ParticleEmitter BoundEmitter
		{
			get { return new ParticleEmitter(this.pParticleEmitter); }
		}
		/// <summary>
		/// Gets the static object that is bound to this slot. Returned object is not valid if the static
		/// object is not bound to this slot.
		/// </summary>
		public StaticObject BoundStaticObject
		{
			get { return new StaticObject(this.pStatObj); }
		}
		/// <summary>
		/// Gets valid object that represents the light source that is bound to this slot. Returned object
		/// is not valid if the light source is not bound to this slot.
		/// </summary>
		public LightSource BoundLight
		{
			get { return new LightSource(this.pLight); }
		}

		/// <summary>
		/// Gets material that is used to render this slot.
		/// </summary>
		public Material Material
		{
			get { return this.pMaterial; }
		}
		#endregion
	}
}

File: MonoInterface/TypeMapChecks/EntitySlotInfo.cpp
#include "stdafx.h"

#include "CheckingBasics.h"

TYPE_MIRROR struct EntitySlotInfo
{
	// Slot flags.
	int nFlags;
	// Index of parent slot, (-1 if no parent)
	int nParentSlot;
	// Hide mask used by breakable object to indicate what index of the CStatObj sub-object is hidden.
	uint64 nSubObjHideMask;
	// Slot local transformation matrix.
	const Matrix34 *pLocalTM;
	// Slot world transformation matrix.
	const Matrix34 *pWorldTM;
	// Objects that can binded to the slot.
	EntityId                   entityId;
	struct IStatObj*           pStatObj;
	struct ICharacterInstance*   pCharacter;
	struct IParticleEmitter*   pParticleEmitter;
	struct ILightSource*      pLight;
	struct IRenderNode*      pChildRenderNode;
	struct IGeomCacheRenderNode* pGeomCacheRenderNode;
	// Custom Material used for the slot.
	IMaterial* pMaterial;

	EntitySlotInfo(SEntitySlotInfo other)
	{
		CHECK_TYPE_SIZE(EntitySlotInfo);

		ASSIGN_FIELD(nFlags);
		ASSIGN_FIELD(nParentSlot);
		ASSIGN_FIELD(nSubObjHideMask);
		ASSIGN_FIELD(pLocalTM);
		ASSIGN_FIELD(pWorldTM);
		ASSIGN_FIELD(entityId);
		ASSIGN_FIELD(pStatObj);
		ASSIGN_FIELD(pCharacter);
		ASSIGN_FIELD(pParticleEmitter);
		ASSIGN_FIELD(pLight);
		ASSIGN_FIELD(pChildRenderNode);
		ASSIGN_FIELD(pGeomCacheRenderNode);
		ASSIGN_FIELD(pMaterial);

		CHECK_TYPE(nFlags);
		CHECK_TYPE(nParentSlot);
		CHECK_TYPE(nSubObjHideMask);
		CHECK_TYPE(pLocalTM);
		CHECK_TYPE(pWorldTM);
		CHECK_TYPE(entityId);
		CHECK_TYPE(pStatObj);
		CHECK_TYPE(pCharacter);
		CHECK_TYPE(pParticleEmitter);
		CHECK_TYPE(pLight);
		CHECK_TYPE(pChildRenderNode);
		CHECK_TYPE(pGeomCacheRenderNode);
		CHECK_TYPE(pMaterial);
	}
};